### PR TITLE
Fix: Type error on `<Image />` component when strict TypeScript rules are used with library type checking.

### PR DESCRIPTION
**Pull request by admirsaheta**
## Changes

- What does this change?
Fixes the following typescript issue, with it now being explicitly defined to avoid type mismatch on the runtime interpreter
```
Element implicitly has an 'any' type because expression of type '"data-image-component"' can't be used to index type 'HTMLAttributes<"img">'.
  Property 'data-image-component' does not exist on type 'HTMLAttributes<"img">'.
```
image

image

### CODE REVIEW

File: .changeset/blue-dolls-melt.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"astro": major
+---
+
+hotfix:explicit-typesafe-fix
```
Comment (Fryuni) [inline .changeset/blue-dolls-melt.md:5]: Fixes `<Image />` type error when strict TypeScript rules are used with library type checking.

**Review by Fryuni — CHANGES_REQUESTED** [review]
Hey @admirsaheta, thanks for the PR! We are always glad to have new contributors.

There are just some things to improve on the changeset:
- Those messages get added to the [release notes](https://github.com/withastro/astro/releases/tag/astro%404.5.9) and to the [changelog](https://github.com/withastro/astro/blob/main/packages/astro/CHANGELOG.md#459), so we use clear descriptions of what was changed or fixed for people to understand how it might affect them.
- The changeset also indicates how the change relates to [SemVer](https://semver.org/), so a "major" there would mean a breaking change that would have to wait until Astro 5 to be included. That is not the case here. Since this PR doesn't change any Astro behavior and fixes a problem it can be a patch.

**Review by Fryuni — APPROVED** [review]

File: .changeset/blue-dolls-melt.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"astro": patch
+---
+
+Fixes <Image /> type error when strict TypeScript rules are used with library type checking.
```
Comment (florian-lefebvre) [inline .changeset/blue-dolls-melt.md:5]:
```suggestion
Fixes `<Image />` type error when strict TypeScript rules are used with library type checking
```

**Comment by Princesseuh** [comment]
Hmm, I think there's a better fix here. `HTMLAttributes` not allowing data attributes is a problem

**Comment by admirsaheta** [comment]
> Hmm, I think there's a better fix here. `HTMLAttributes` not allowing data attributes is a problem

You could extend the interface and provide custom attributes there e.g 

```ts
interface AdditionalAttributes extends HTMLAttributes<HTMLImageElement> {
    'data-image-component'?: string;
}
```
Let me know which approach fits better

**Comment by florian-lefebvre** [comment]
We have been discussing on discord to update the `HTMLAttributes`. So at https://github.com/withastro/astro/blob/1cd2a740221ee14267f2889c4eb200bbcecb08aa/packages/astro/types.d.ts#L8 allow such type:
```ts
[key: `${string}-${string}`]: any
```
Do you want to handle it?

**Comment by florian-lefebvre** [comment]
There's no rush don't worry!

**Comment by admirsaheta** [comment]
Should handle it correctly here now 👍

File: packages/astro/types.d.ts [hunk 3]
```diff
@@ -4,11 +4,14 @@ import type { Simplify } from './dist/type-utils.js';
...
 export type HTMLTag = keyof astroHTML.JSX.DefinedIntrinsicElements;
+
 /** The built-in attributes for any known HTML or SVG element name */
 export type HTMLAttributes<Tag extends HTMLTag> = Omit<
 	astroHTML.JSX.IntrinsicElements[Tag],
-	keyof Omit<AstroBuiltinAttributes, 'class:list'>
->;
+	keyof Omit<AstroBuiltinAttributes, 'class:list'> 
+> & {
+	[key: string]: string | number | boolean | undefined;
```
Comment (florian-lefebvre) [inline packages/astro/types.d.ts:13]: @Princesseuh Do you prefer `any`?

File: .changeset/blue-dolls-melt.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"astro": patch
+---
+
+Fixes `<Image />` type error when strict TypeScript rules are used with library type checking
```
Comment (florian-lefebvre) [inline .changeset/blue-dolls-melt.md:5]:
```suggestion
Updated the `HTMLAttributes` type exported from `astro` to allow data attributes
```

**Review by florian-lefebvre — APPROVED** [review]
LGTM

File: packages/astro/types.d.ts (hunk 3 above)
Comment (Princesseuh) [reply packages/astro/types.d.ts:13]: Without necessarily using `any`, allowing also `null` would be great

Comment (admirsaheta) [reply packages/astro/types.d.ts]: > Without necessarily using `any`, allowing also `null` would be great

Done

**Review by Princesseuh — APPROVED** [review]

**Pull request merged**